tests/model_inheritance/test_abstract_inheritance.py

```diff
@@ -1,64 +1,184 @@
+
+    def test_derived_from_descendant_ono(self):
+        self.assertEqual(AbstractInheritanceSingleParentTests.DerivedFromDescendantONO._meta.get_field('foo').max_length, 40)
+
+    def test_derived_from_descendant_oo(self):
+        self.assertEqual(AbstractInheritanceSingleParentTests.DerivedFromDescendantOO._meta.get_field('foo').max_length, 50)
+
+
+class AbstractInheritanceMultiParentTests(SimpleTestCase):
+    """ Test mro preserved when inheriting from multiple parents at once.
+        NO = non-overriding, O = overriding.
+    """
+    class AbstractBase1(models.Model):
+        foo = models.CharField(max_length=30)
+
+        class Meta:
+            abstract = True
+
+    class AbstractBase2(models.Model):
+        foo = models.CharField(max_length=40)
+
+        class Meta:
+            abstract = True
+
+    class AbstractBase3(models.Model):
+        class Meta:
+            abstract = True
+
+    class DerivedFromBases12NO(AbstractBase1, AbstractBase2):
+        pass
+
+    class DerivedFromBases13NO(AbstractBase1, AbstractBase3):
+        pass
+
+    class DerivedFromBases32NO(AbstractBase3, AbstractBase2):
+        pass
+
+    class DerivedFromBases12O(AbstractBase1, AbstractBase2):
+        foo = models.CharField(max_length=50)
+
+    class DerivedFromBases13O(AbstractBase1, AbstractBase3):
+        foo = models.CharField(max_length=50)
+
+    class DerivedFromBases32O(AbstractBase3, AbstractBase2):
+        foo = models.CharField(max_length=50)
+
+    def test_derived_from_bases_12no(self):
+        self.assertEqual(AbstractInheritanceMultiParentTests.DerivedFromBases12NO._meta.get_field('foo').max_length, 30)
 
+    def test_derived_from_bases_13no(self):
+        self.assertEqual(AbstractInheritanceMultiParentTests.DerivedFromBases13NO._meta.get_field('foo').max_length, 30)
+
+    def test_derived_from_bases_32no(self):
+        self.assertEqual(AbstractInheritanceMultiParentTests.DerivedFromBases32NO._meta.get_field('foo').max_length, 40)
+
+    def test_derived_from_bases_12o(self):
+        self.assertEqual(AbstractInheritanceMultiParentTests.DerivedFromBases12O._meta.get_field('foo').max_length, 50)
+
+    def test_derived_from_bases_13o(self):
+        self.assertEqual(AbstractInheritanceMultiParentTests.DerivedFromBases13O._meta.get_field('foo').max_length, 50)
+
+    def test_derived_from_bases_32o(self):
+        self.assertEqual(AbstractInheritanceMultiParentTests.DerivedFromBases32O._meta.get_field('foo').max_length, 50)
+
+
+class AbstractInheritanceMultiParentDiamondTests(SimpleTestCase):
+    """ Test mro preserved when inheriting from multiple parents in diamond formation.
+        NO = non-overriding, O = overriding.
+
+        NOTE: We need a rewrite of `ModelBase.__new__()` to support this properly.
+        Until that happens, one of the tests here is marked as expected failure. Startup checks will
+        attempt to prevent users from building these cases anyway.
```

Comment:
@carltongibson I've updated this note in response to @adamchainz' feedback, as well given what you said about the diamond check above. I think it's correct now and useful (?) to explain why we've marked one of the tests with `@expectedfailure`, but can remove it altogether if it still feels more like a ticket that's been shoved into the code.